runtime/memprof.c

```diff
@@ -0,0 +1,316 @@
+   less than 10^-6, which is way better than what we need.
+ */
+static double lfact(double n) {
+  static double tab[10] = {
+    0.0000000000, 0.0000000000, 0.6931471806, 1.7917594692, 3.1780538303,
+    4.7874917428, 6.5792512120, 8.5251613611, 10.6046029027, 12.8018274801 };
+  if(n < 10)
+    return tab[(int)n];
+  return n*(log(n) - 1) + (1./6)*log(((8*n + 4)*n + 1)*n + 1./34)
+       + 0.5723649429; /* log(pi)/2 */
+}
+
+static double next_mt_generate_poisson;
+/* Max returned value : 2^30-2. Assumes lambda >= 0. */
+static int32_t mt_generate_poisson(double len) {
+  double cur_lambda = lambda * len;
+  CAMLassert(cur_lambda >= 0 && cur_lambda < 1e20);
+
+  if(caml_memprof_suspended || cur_lambda == 0)
+    return 0;
+
+  if(cur_lambda < 20) {
+    /* First algorithm when [cur_lambda] is small: we proceed by
+       repeated simulations of exponential distributions. */
+
+    next_mt_generate_poisson -= cur_lambda;
+    if(next_mt_generate_poisson > 0) {
+      /* Fast path if [cur_lambda] is small: we reuse the same
+         exponential sample accross several calls to
+         [mt_generate_poisson]. */
+      return 0;
+    } else {
+      /* We use the float versions of exp/log, since these functions
+         are significantly faster, and we really don't need much
+         precision here. The entropy contained in
+         [next_mt_generate_poisson] is anyway bounded by the entropy
+         provided by [mt_generate_uniform], which is 32bits. */
+      double p = expf(-next_mt_generate_poisson);
+      int32_t k = 0;
+      do {
+        k++;
+        p *= mt_generate_uniform();
+      } while(p > 1);
+
+      /* [p] is now uniformly distributed in [0, 1] and independent
+         from other variables (including [k]). We can therefore reuse
+         [p] for reinitializing [next_mt_generate_poisson]. */
+      next_mt_generate_poisson = -logf(p);
+
+      return k;
+    }
+
+  } else {
+
+    /* Second algorithm when [cur_lambda] is large. Taken from: */
+    /* The Computer Generation of Poisson Random Variables
+       A. C. Atkinson Journal of the Royal Statistical Society.
+       Series C (Applied Statistics) Vol. 28, No. 1 (1979), pp. 29-35
+       "Method PA" */
+
+    double c, beta_rec, k, log_cur_lambda;
+    log_cur_lambda = log(cur_lambda);
+    c = 0.767 - 3.36/cur_lambda;
+    beta_rec = sqrt((3./(PI*PI))*cur_lambda);
+    k = log(c*beta_rec) - cur_lambda;
+    while(1) {
+      double u, n, v, y;
+      u = mt_generate_uniform();
+      y = log(1./u-1);
+      n = floor(cur_lambda - y*beta_rec + 0.5);
+      if(n < 0.)
+        continue;
+      v = mt_generate_uniform();
+      /* When [cur_lambda] is large, we expect [n*log_cur_lambda] and
+         [lfact(n)] to be close, while both being relatively
+         large. Hence, here, we may actually need the double precision
+         in the computation of log and lfact. */
+      if(y + log(v*u*u) < k + n*log_cur_lambda - lfact(n))
```

Comment:
In the secondary source, the check is

```
	lhs = y + log(v/(1.0 + exp(y))^2)
	rhs = k + n*log(lambda) - log(n!)
	if (lhs <= rhs)
		return n
```

Note:
- @jhjourdan uses a strict equality rather than a large equality. I don't think this makes any difference.
- in @jhjourdan's code, `n` is clipped in the return clause to be at most 2^30-2. Why not 2^30-1, the largest representable int32_t?.
- The right-hand-side of the comparisons are exactly identical.
- On the other hand, the left-hand-side differ, but they are in fact equivalent. Indeed, the secondary source uses `y + log(v/(1.0 + exp(y))^2)`, but `y = log(1/u - 1)` gives `log(v/(1+exp(y))^2) = log(v * (1 + 1/u - 1)^2) = log(v*u*u)`.

Reply:
Note that, however, this higher bound should not be reached with any reasonable use of statmemprof.

Reply:
Thanks for the clarification. Could the output interval of the function be documented somewhere, for example as a comment on its prototype? (If you want to leave some wiggle room for future computations, maybe it would be sensible to clip to 2^29 instead of the very use-case-specific 2^30-2?).

Reply:
Done.